<a href="https://colab.research.google.com/github/surendarmanoj/data-science/blob/main/Exploratory_Data_Analysis_Zomato_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Zomato Dataset -> [Dataset](https://drive.google.com/drive/folders/1F1nT78Yet2BrR7LNwkSBi6Px-ggzane_?usp=sharing)

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import json

In [22]:
df = pd.read_csv("merged_file.csv")
df

has_online_delivery                                         photos_url  \
0                        1  https://www.zomato.com/HauzKhasSocial/photos?u...   
1                        0  https://www.zomato.com/ncr/qubitos-the-terrace...   
2                        1  https://www.zomato.com/ncr/the-hudson-cafe-del...   
3                        0  https://www.zomato.com/ncr/summer-house-cafe-h...   
4                        0  https://www.zomato.com/ncr/38-barracks-connaug...   
...                    ...                                                ...   
29748                    0  https://www.zomato.com/pocatello-id/senor-igua...   
29749                    0  https://www.zomato.com/pocatello-id/sandpiper-...   
29750                    0  https://www.zomato.com/pocatello-id/texas-road...   
29751                    0  https://www.zomato.com/pocatello-id/riverwalk-...   
29752                    0  https://www.zomato.com/pocatello-id/royal-hote...   

                                                     url  price_range  \
0      https://www.zomato.com/HauzKhasSocial?utm_sour...            3   
1      https://www.zomato.com/ncr/qubitos-the-terrace...            3   
2      https://www.zomato.com/ncr/the-hudson-cafe-del...            2   
3      https://www.zomato.com/ncr/summer-house-cafe-h...            3   
4      https://www.zomato.com/ncr/38-barracks-connaug...            3   
...                                                  ...          ...   
29748  https://www.zomato.com/pocatello-id/senor-igua...            1   
29749  https://www.zomato.com/pocatello-id/sandpiper-...            3   
29750  https://www.zomato.com/pocatello-id/texas-road...            3   
29751  https://www.zomato.com/pocatello-id/riverwalk-...            1   
29752  https://www.zomato.com/pocatello-id/royal-hote...            1   

                                 apikey                           name  \
0      b90e6a8c738410315a20c449fe2eb1b1               Hauz Khas Social   
1      b90e6a8c738410315a20c449fe2eb1b1     Qubitos - The Terrace Cafe   
2      b90e6a8c738410315a20c449fe2eb1b1                The Hudson Cafe   
3      b90e6a8c738410315a20c449fe2eb1b1              Summer House Cafe   
4      b90e6a8c738410315a20c449fe2eb1b1                    38 Barracks   
...                                 ...                            ...   
29748  b90e6a8c738410315a20c449fe2eb1b1                  Senor Iguanas   
29749  b90e6a8c738410315a20c449fe2eb1b1  Sandpiper Restaurant & Lounge   
29750  b90e6a8c738410315a20c449fe2eb1b1                Texas Roadhouse   
29751  b90e6a8c738410315a20c449fe2eb1b1                 Riverwalk Cafe   
29752  b90e6a8c738410315a20c449fe2eb1b1                    Royal Hotel   

                                                cuisines  is_delivering_now  \
0             Continental, American, Asian, North Indian                  0   
1      Thai, European, Mexican, North Indian, Chinese...                  0   
2                    Cafe, Italian, Continental, Chinese                  0   
3                                   Italian, Continental                  0   
4                 North Indian, Italian, Asian, American                  0   
...                                                  ...                ...   
29748                                            Mexican                  0   
29749                           American, Seafood, Steak                  0   
29750                               American, BBQ, Steak                  0   
29751                                        Asian, Thai                  0   
29752                                    Pizza, Bar Food                  0   

                           deeplink  \
0        zomato://restaurant/308322   
1      zomato://restaurant/18037817   
2        zomato://restaurant/312345   
3        zomato://restaurant/307490   
4      zomato://restaurant/18241537   
...                             ...   
29748  zomato://restaurant/17582627   
29749  

In [23]:
df_default = pd.read_csv("zomato.csv", encoding="latin")

In [24]:
df_default

Restaurant ID           Restaurant Name  Country Code              City  \
0           6317637          Le Petit Souffle           162       Makati City   
1           6304287          Izakaya Kikufuji           162       Makati City   
2           6300002    Heat - Edsa Shangri-La           162  Mandaluyong City   
3           6318506                      Ooma           162  Mandaluyong City   
4           6314302               Sambo Kojin           162  Mandaluyong City   
...             ...                       ...           ...               ...   
9546        5915730              NamlÛ± Gurme           208         ÛÁstanbul   
9547        5908749             Ceviz AÛôacÛ±           208         ÛÁstanbul   
9548        5915807                     Huqqa           208         ÛÁstanbul   
9549        5916112              Aôôk Kahve           208         ÛÁstanbul   
9550        5927402  Walter's Coffee Roastery           208         ÛÁstanbul   

                                                Address  \
0     Third Floor, Century City Mall, Kalayaan Avenu...   
1     Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2     Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3     Third Floor, Mega Fashion Hall, SM Megamall, O...   
4     Third Floor, Mega Atrium, SM Megamall, Ortigas...   
...                                                 ...   
9546  Kemankeô Karamustafa Paôa Mahallesi, RÛ±htÛ±...   
9547  Koôuyolu Mahallesi, Muhittin íìstí_ndaÛô Cadd...   
9548  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9549  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9550  CafeaÛôa Mahallesi, BademaltÛ± Sokak, No 21/B,...   

                                        Locality  \
0      Century City Mall, Poblacion, Makati City   
1     Little Tokyo, Legaspi Village, Makati City   
2     Edsa Shangri-La, Ortigas, Mandaluyong City   
3         SM Megamall, Ortigas, Mandaluyong City   
4         SM Megamall, Ortigas, Mandaluyong City   
...                                          ...   
9546                                    Karakí_y   
9547                                   Koôuyolu   
9548                                 Kuruí_eôme   
9549                                 Kuruí_eôme   
9550                                        Moda   

                                       Locality Verbose   Longitude  \
0     Century City Mall, Poblacion, Makati City, Mak...  121.027535   
1     Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101   
2     Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831   
3     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475   
4     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508   
...                                                 ...         ...   
9546                                Karakí_y, ÛÁstanbul   28.977392   
9547                               Koôuyolu, ÛÁstanbul   29.041297   
9548                             Kuruí_eôme, ÛÁstanbul   29.034640   
9549                             Kuruí_eôme, ÛÁstanbul   29.036019   
9550                                    Moda, ÛÁstanbul   29.026016   

       Latitude                          Cuisines  ...          Currency  \
0     14.565443        French, Japanese, Desserts  ...  Botswana Pula(P)   
1     14.553708                          Japanese  ...  Botswana Pula(P)   
2     14.581404  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)   
3     14.585318                   Japanese, Sushi  ...  Botswana Pula(P)   
4     14.584450                  Japanese, Korean  ...  Botswana Pula(P)   
...         ...                               ...  ...               ...   
9546  41.022793                           Turkish  ...  Turkish Lira(TL)   
9547  41.009847   World Cuisine, Patisserie, Cafe  ...  Turkish Lira(TL)   
9548  41.055817            Italian, World Cuisine  ...  Turkish Lira(TL)   
9549  41.057979                   Restaurant Cafe  ...  Turkish Lira(TL)   
9550  40.9

In [25]:
df1 = df.copy()

In [26]:
df1.columns

Index(['has_online_delivery', 'photos_url', 'url', 'price_range', 'apikey',
       'name', 'cuisines', 'is_delivering_now', 'deeplink', 'menu_url',
       'average_cost_for_two', 'book_url', 'switch_to_order_menu', 'offers',
       'has_table_booking', 'featured_image', 'zomato_events', 'currency',
       'id', 'thumb', 'establishment_types', 'events_url',
       'user_rating.rating_text', 'user_rating.rating_color',
       'user_rating.votes', 'user_rating.aggregate_rating', 'R.res_id',
       'location.latitude', 'location.address', 'location.city',
       'location.country_id', 'location.locality_verbose', 'location.city_id',
       'location.zipcode', 'location.longitude', 'location.locality',
       'order_deeplink', 'order_url'],
      dtype='object')

In [27]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29753 entries, 0 to 29752
Data columns (total 38 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   has_online_delivery           29753 non-null  int64  
 1   photos_url                    29753 non-null  object 
 2   url                           29753 non-null  object 
 3   price_range                   29753 non-null  int64  
 4   apikey                        29753 non-null  object 
 5   name                          29753 non-null  object 
 6   cuisines                      29563 non-null  object 
 7   is_delivering_now             29753 non-null  int64  
 8   deeplink                      29753 non-null  object 
 9   menu_url                      29753 non-null  object 
 10  average_cost_for_two          29753 non-null  int64  
 11  book_url                      4291 non-null   object 
 12  switch_to_order_menu          29753 non-null  int64  
 13  o

This Zomato dataset was compiled from JSON files containing response details obtained via API calls to the server. The dataset consists of 29,753 records and 38 columns, with 23 qualitative variables and the rest quantitative.

Upon further investigation, we identified that the orders column contains a list of values. We will convert this into a boolean format: if the list is empty, it will be set to 0 (False), otherwise it will be marked as 1 (True). The same action will be applied to the zomato_events column.

Additionally, we observed that the aggregate_rating and rating_color columns exhibit identical behavior, which will be further analyzed during the exploratory data analysis (EDA) phase.

In [28]:
df1.describe()

has_online_delivery   price_range  is_delivering_now  \
count         29753.000000  29753.000000       29753.000000   
mean              0.192115      2.183444           0.001949   
std               0.393970      0.954264           0.044110   
min               0.000000      1.000000           0.000000   
25%               0.000000      1.000000           0.000000   
50%               0.000000      2.000000           0.000000   
75%               0.000000      3.000000           0.000000   
max               1.000000      4.000000           1.000000   

       average_cost_for_two  switch_to_order_menu  has_table_booking  \
count           29753.00000               29753.0       29753.000000   
mean             2544.60888                   0.0           0.144221   
std             18230.51240                   0.0           0.351319   
min                 0.00000                   0.0           0.000000   
25%                50.00000                   0.0           0.000000   
50%               450.00000                   0.0           0.000000   
75%              1000.00000                   0.0           0.000000   
max            800000.00000                   0.0           1.000000   

                 id  user_rating.votes  user_rating.aggregate_rating  \
count  2.975300e+04       29753.000000                  29753.000000   
mean   1.035475e+07         522.612274                      3.325201   
std    8.534506e+06         948.450729                      1.270941   
min    4.000000e+00           0.000000                      0.000000   
25%    3.064070e+05          19.000000                      3.100000   
50%    1.660886e+07          89.000000                      3.700000   
75%    1.824528e+07         542.000000                      4.100000   
max    1.850065e+07       10934.000000                      4.900000   

           R.res_id  location.latitude  location.country_id  location.city_id  \
count  2.975300e+04       29753.000000         29753.000000      29753.000000   
mean   1.035475e+07          19.114803            32.271401       1255.543374   
std    8.534506e+06          23.496187            69.626147       2752.564946   
min    4.000000e+00         -41.330428             1.000000          1.000000   
25%    3.064070e+05          23.044337             1.000000          1.000000   
50%    1.660886e+07          28.567720             1.000000          1.000000   
75%    1.824528e+07          28.642216            14.000000       1353.000000   
max    1.850065e+07          73.990000           216.000000      11071.000000   

       location.longitude  order_deeplink  
count        29753.000000             0.0  
mean            63.466626             NaN  
std             68.264524             NaN  
min           -157.948486             NaN  
25%             77.088328             NaN  
50%             77.219739             NaN  
75%             77.363084             NaN  
max            174.832089             NaN

In [29]:
df1.isnull().sum()

has_online_delivery                 0
photos_url                          0
url                                 0
price_range                         0
apikey                              0
name                                0
cuisines                          190
is_delivering_now                   0
deeplink                            0
menu_url                            0
average_cost_for_two                0
book_url                        25462
switch_to_order_menu                0
offers                              0
has_table_booking                   0
featured_image                   8375
zomato_events                   27918
currency                           20
id                                  0
thumb                            8375
establishment_types                 0
events_url                          0
user_rating.rating_text             0
user_rating.rating_color            0
user_rating.votes                   0
user_rating.aggregate_rating        0
R.res_id                            0
location.latitude                   0
location.address                    0
location.city                       0
location.country_id                 0
location.locality_verbose           0
location.city_id                    0
location.zipcode                14541
location.longitude                  0
location.locality                   0
order_deeplink                  29753
order_url                       29695
dtype: int64

In [30]:
df2 = df1[['has_online_delivery', 'price_range',
       'name', 'cuisines', 'is_delivering_now',
       'average_cost_for_two', 'switch_to_order_menu', 'offers',
       'has_table_booking','zomato_events', 'currency', 'establishment_types',
       'user_rating.rating_text',
       'user_rating.votes', 'user_rating.aggregate_rating','location.address', 'location.city',
       'location.country_id', 'location.locality_verbose']]

df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29753 entries, 0 to 29752
Data columns (total 19 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   has_online_delivery           29753 non-null  int64  
 1   price_range                   29753 non-null  int64  
 2   name                          29753 non-null  object 
 3   cuisines                      29563 non-null  object 
 4   is_delivering_now             29753 non-null  int64  
 5   average_cost_for_two          29753 non-null  int64  
 6   switch_to_order_menu          29753 non-null  int64  
 7   offers                        29753 non-null  object 
 8   has_table_booking             29753 non-null  int64  
 9   zomato_events                 1835 non-null   object 
 10  currency                      29733 non-null  object 
 11  establishment_types           29753 non-null  object 
 12  user_rating.rating_text       29753 non-null  object 
 13  u

In our exploratory data analysis (EDA), we refined the original dataset by selecting features relevant to our analysis and dropping those that were not necessary. The original dataset contained a total of 38 columns, but we have created a new dataframe df2 with 19 columns to streamline our analysis.





| Column Name                      | Data Type | Description                                                                                      |
|----------------------------------|-----------|--------------------------------------------------------------------------------------------------|
| **has_online_delivery**          | int64     | Indicates whether the restaurant provides online delivery (`1`) or not (`0`).                    |
| **price_range**                  | int64     | Price range category of the restaurant (e.g., `1` for low, up to `4` for high).                  |
| **name**                         | object    | Name of the restaurant.                                                                          |
| **cuisines**                     | object    | Types of cuisines offered; multiple cuisines are comma-separated.                                |
| **is_delivering_now**            | int64     | Indicates if the restaurant is currently delivering (`1`) or not (`0`).                          |
| **average_cost_for_two**         | int64     | Average cost for two people dining at the restaurant.                                            |
| **switch_to_order_menu**         | int64     | Binary flag indicating a menu switch for ordering.                                               |
| **offers**                       | object    | Details of any current offers; can be empty or contain offer information.                        |
| **has_table_booking**            | int64     | Indicates whether the restaurant accepts table bookings (`1`) or not (`0`).                      |
| **zomato_events**                | object    | Information on events related to the restaurant; many entries are null.                          |
| **currency**                     | object    | Currency in which prices are listed.                                                             |
| **establishment_types**          | object    | Type of establishment (e.g., Café, Fine Dining).                                                 |
| **user_rating.rating_text**      | object    | Textual representation of the user rating (e.g., 'Excellent').                                   |
| **user_rating.votes**            | int64     | Number of votes contributing to the user rating.                                                 |
| **user_rating.aggregate_rating** | float64   | Aggregate user rating out of 5.                                                                  |
| **location.address**             | object    | Physical address of the restaurant.                                                              |
| **location.city**                | object    | City where the restaurant is located.                                                            |
| **location.country_id**          | int64     | Numeric ID representing the country.                                                             |
| **location.locality_verbose**    | object    | Detailed locality information, often including neighborhood and city.                            |




**Rationale for Dropping Features:**

The original dataset included additional features that were not essential for our analysis, such as:

**URLs** ('photos_url', 'url', 'deeplink', 'menu_url', 'book_url', 'featured_image', 'events_url', 'order_deeplink', 'order_url')

**API keys and IDs** ('apikey', 'id', 'R.res_id')

**Location coordinates** ('location.latitude', 'location.longitude', 'location.city_id', 'location.zipcode', 'location.locality')

**Redundant or less informative fields** ('thumb', 'user_rating.rating_color')
By excluding these features, we focused on the most relevant data, simplifying the dataset for more effective analysis.

In [31]:
df3 = {f: df2[f].isnull().sum() for f in df2.columns if df2[f].isnull().sum() > 0}
df3

{'cuisines': 190, 'zomato_events': 27918, 'currency': 20}

In [32]:
df2 = df2.dropna(subset=['cuisines'])
df2.reset_index(drop=True, inplace=True)

In [33]:
df3 = {f: df2[f].isnull().sum() for f in df2.columns if df2[f].isnull().sum() > 0}
df3

{'zomato_events': 27728, 'currency': 20}

In [34]:
df2 = df2.dropna(subset=['currency'])
df2.reset_index(drop=True, inplace=True)

In [35]:
df3 = {f: df2[f].isnull().sum() for f in df2.columns if df2[f].isnull().sum() > 0}
df3

{'zomato_events': 27708}

In [36]:
df2['zomato_events'] = df2['zomato_events'].notnull().astype(int)

<ipython-input-36-57f456337630>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['zomato_events'] = df2['zomato_events'].notnull().astype(int)


In [37]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29543 entries, 0 to 29542
Data columns (total 19 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   has_online_delivery           29543 non-null  int64  
 1   price_range                   29543 non-null  int64  
 2   name                          29543 non-null  object 
 3   cuisines                      29543 non-null  object 
 4   is_delivering_now             29543 non-null  int64  
 5   average_cost_for_two          29543 non-null  int64  
 6   switch_to_order_menu          29543 non-null  int64  
 7   offers                        29543 non-null  object 
 8   has_table_booking             29543 non-null  int64  
 9   zomato_events                 29543 non-null  int64  
 10  currency                      29543 non-null  object 
 11  establishment_types           29543 non-null  object 
 12  user_rating.rating_text       29543 non-null  object 
 13  u

In [38]:
df2.head()

has_online_delivery  price_range                        name  \
0                    1            3            Hauz Khas Social   
1                    0            3  Qubitos - The Terrace Cafe   
2                    1            2             The Hudson Cafe   
3                    0            3           Summer House Cafe   
4                    0            3                 38 Barracks   

                                            cuisines  is_delivering_now  \
0         Continental, American, Asian, North Indian                  0   
1  Thai, European, Mexican, North Indian, Chinese...                  0   
2                Cafe, Italian, Continental, Chinese                  0   
3                               Italian, Continental                  0   
4             North Indian, Italian, Asian, American                  0   

   average_cost_for_two  switch_to_order_menu offers  has_table_booking  \
0                  1600                     0     []                  1   
1                  1500                     0     []                  1   
2                   850                     0     []                  0   
3                  1850                     0     []                  1   
4                  1600                     0     []                  1   

   zomato_events currency establishment_types user_rating.rating_text  \
0              1      Rs.                  []               Very Good   
1              1      Rs.                  []               Excellent   
2              0      Rs.                  []               Very Good   
3              1      Rs.                  []               Very Good   
4              0      Rs.                  []               Very Good   

   user_rating.votes  user_rating.aggregate_rating  \
0               7931                           4.3   
1                778                           4.5   
2               1537                           4.4   
3               1823                           4.1   
4                840                           4.4   

                                    location.address location.city  \
0             9-A & 12, Hauz Khas Village, New Delhi     New Delhi   
1  C-7, Vishal Enclave, Opposite Metro Pillar 417...     New Delhi   
2  2524, 1st Floor, Hudson Lane, Delhi University...     New Delhi   
3  1st Floor, DDA Shopping Complex, Aurobindo Pla...     New Delhi   
4     M-38, Outer Circle, Connaught Place, New Delhi     New Delhi   

   location.country_id              location.locality_verbose  
0                    1           Hauz Khas Village, New Delhi  
1                    1              Rajouri Garden, New Delhi  
2                    1  Delhi University-GTB Nagar, New Delhi  
3                    1                   Hauz Khas, New Delhi  
4                    1             Connaught Place, New Delhi

In [53]:
df5 = df1[['has_online_delivery', 'price_range',
           'name', 'cuisines', 'is_delivering_now',
           'average_cost_for_two', 'switch_to_order_menu', 'offers',
           'has_table_booking', 'zomato_events', 'currency',
           'user_rating.rating_text',
           'user_rating.votes', 'user_rating.aggregate_rating', 'location.address', 'location.city',
           'location.country_id', 'location.locality_verbose']]
df5

has_online_delivery  price_range                           name  \
0                        1            3               Hauz Khas Social   
1                        0            3     Qubitos - The Terrace Cafe   
2                        1            2                The Hudson Cafe   
3                        0            3              Summer House Cafe   
4                        0            3                    38 Barracks   
...                    ...          ...                            ...   
29748                    0            1                  Senor Iguanas   
29749                    0            3  Sandpiper Restaurant & Lounge   
29750                    0            3                Texas Roadhouse   
29751                    0            1                 Riverwalk Cafe   
29752                    0            1                    Royal Hotel   

                                                cuisines  is_delivering_now  \
0             Continental, American, Asian, North Indian                  0   
1      Thai, European, Mexican, North Indian, Chinese...                  0   
2                    Cafe, Italian, Continental, Chinese                  0   
3                                   Italian, Continental                  0   
4                 North Indian, Italian, Asian, American                  0   
...                                                  ...                ...   
29748                                            Mexican                  0   
29749                           American, Seafood, Steak                  0   
29750                               American, BBQ, Steak                  0   
29751                                        Asian, Thai                  0   
29752                                    Pizza, Bar Food                  0   

       average_cost_for_two  switch_to_order_menu offers  has_table_booking  \
0                      1600                     0     []                  1   
1                      1500                     0     []                  1   
2                       850                     0     []                  0   
3                      1850                     0     []                  1   
4                      1600                     0     []                  1   
...                     ...                   ...    ...                ...   
29748                     0                     0     []                  0   
29749                    40                     0     []                  0   
29750                    45                     0     []                  0   
29751                    10                     0     []                  0   
29752                     0                     0     []                  0   

                                           zomato_events currency  \
0      [{'event': {'display_date': '05 April - 21 May...      Rs.   
1      [{'event': {'display_date': '16 April - 17 Apr...      Rs.   
2                                                    NaN      Rs.   
3      [{'event': {'display_date': '16 April - 17 Apr...      Rs.   
4                                                    NaN      Rs.   
...                                                  ...      ...   
29748                                                NaN        $   
29749                                                NaN        $   
29750                                                NaN        $   
29751                                                NaN        $   
29752                                                NaN        $   

      user_rating.rating_text  user_rating.votes  \
0                   Very Good               7931   
1                   Excellent                778   
2                   Very Good               1537   
3                   Very Good               1823   
4                   Very Good                840   
...                       ...                ...   
29748                    Go

In [54]:
column_mapping = {
    'has_online_delivery':'Online_Available',
    'price_range':"Average_Price",
    'name':'Name',
    'cuisines':'Menu',
    'is_delivering_now':"Open",
    'average_cost_for_two':"Cost_per_table",
    'offers':'Offers',
    'has_table_booking':'Booking_available',
    'zomato_events':'Events',
    'currency':'Currency',
    'user_rating.rating_text':"Rating_text",
    'user_rating.votes':'Votes',
    'user_rating.aggregate_rating':"Aggregate_Rating",
    'location.address':'Address',
    'location.city':'City',
    'location.country_id':'C_ID',
    'location.locality_verbose':'Neighbourhood'
}

# Apply renaming
df5 = df5.rename(columns=column_mapping)

In [55]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29753 entries, 0 to 29752
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Online_Available      29753 non-null  int64  
 1   Average_Price         29753 non-null  int64  
 2   Name                  29753 non-null  object 
 3   Menu                  29563 non-null  object 
 4   Open                  29753 non-null  int64  
 5   Cost_per_table        29753 non-null  int64  
 6   switch_to_order_menu  29753 non-null  int64  
 7   Offers                29753 non-null  object 
 8   Booking_available     29753 non-null  int64  
 9   Events                1835 non-null   object 
 10  Currency              29733 non-null  object 
 11  Rating_text           29753 non-null  object 
 12  Votes                 29753 non-null  int64  
 13  Aggregate_Rating      29753 non-null  float64
 14  Address               29753 non-null  object 
 15  City               

In [56]:
t_list = df5["Offers"]
t_list

0        []
1        []
2        []
3        []
4        []
         ..
29748    []
29749    []
29750    []
29751    []
29752    []
Name: Offers, Length: 29753, dtype: object

In [57]:
t_list = df5["Offers"]

# Transform the list: empty lists to 0, non-empty lists to 1
y_list = [
    0 if isinstance(x, str) and len(x) > 1 and x[1] != "{" else 1 if isinstance(x, str) else 0
    for x in t_list
]

# Optionally, assign the transformed t_list back to the DataFrame
df5["Offers"] = t_list

# Verify the result
df5.head()

Online_Available  Average_Price                        Name  \
0                 1              3            Hauz Khas Social   
1                 0              3  Qubitos - The Terrace Cafe   
2                 1              2             The Hudson Cafe   
3                 0              3           Summer House Cafe   
4                 0              3                 38 Barracks   

                                                Menu  Open  Cost_per_table  \
0         Continental, American, Asian, North Indian     0            1600   
1  Thai, European, Mexican, North Indian, Chinese...     0            1500   
2                Cafe, Italian, Continental, Chinese     0             850   
3                               Italian, Continental     0            1850   
4             North Indian, Italian, Asian, American     0            1600   

   switch_to_order_menu Offers  Booking_available  \
0                     0     []                  1   
1                     0     []                  1   
2                     0     []                  0   
3                     0     []                  1   
4                     0     []                  1   

                                              Events Currency Rating_text  \
0  [{'event': {'display_date': '05 April - 21 May...      Rs.   Very Good   
1  [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Excellent   
2                                                NaN      Rs.   Very Good   
3  [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Very Good   
4                                                NaN      Rs.   Very Good   

   Votes  Aggregate_Rating                                            Address  \
0   7931               4.3             9-A & 12, Hauz Khas Village, New Delhi   
1    778               4.5  C-7, Vishal Enclave, Opposite Metro Pillar 417...   
2   1537               4.4  2524, 1st Floor, Hudson Lane, Delhi University...   
3   1823               4.1  1st Floor, DDA Shopping Complex, Aurobindo Pla...   
4    840               4.4     M-38, Outer Circle, Connaught Place, New Delhi   

        City  C_ID                          Neighbourhood  
0  New Delhi     1           Hauz Khas Village, New Delhi  
1  New Delhi     1              Rajouri Garden, New Delhi  
2  New Delhi     1  Delhi University-GTB Nagar, New Delhi  
3  New Delhi     1                   Hauz Khas, New Delhi  
4  New Delhi     1             Connaught Place, New Delhi

In [59]:
df5.head()

Online_Available  Average_Price                        Name  \
0                 1              3            Hauz Khas Social   
1                 0              3  Qubitos - The Terrace Cafe   
2                 1              2             The Hudson Cafe   
3                 0              3           Summer House Cafe   
4                 0              3                 38 Barracks   

                                                Menu  Open  Cost_per_table  \
0         Continental, American, Asian, North Indian     0            1600   
1  Thai, European, Mexican, North Indian, Chinese...     0            1500   
2                Cafe, Italian, Continental, Chinese     0             850   
3                               Italian, Continental     0            1850   
4             North Indian, Italian, Asian, American     0            1600   

   switch_to_order_menu Offers  Booking_available  \
0                     0     []                  1   
1                     0     []                  1   
2                     0     []                  0   
3                     0     []                  1   
4                     0     []                  1   

                                              Events Currency Rating_text  \
0  [{'event': {'display_date': '05 April - 21 May...      Rs.   Very Good   
1  [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Excellent   
2                                                NaN      Rs.   Very Good   
3  [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Very Good   
4                                                NaN      Rs.   Very Good   

   Votes  Aggregate_Rating                                            Address  \
0   7931               4.3             9-A & 12, Hauz Khas Village, New Delhi   
1    778               4.5  C-7, Vishal Enclave, Opposite Metro Pillar 417...   
2   1537               4.4  2524, 1st Floor, Hudson Lane, Delhi University...   
3   1823               4.1  1st Floor, DDA Shopping Complex, Aurobindo Pla...   
4    840               4.4     M-38, Outer Circle, Connaught Place, New Delhi   

        City  C_ID                          Neighbourhood  
0  New Delhi     1           Hauz Khas Village, New Delhi  
1  New Delhi     1              Rajouri Garden, New Delhi  
2  New Delhi     1  Delhi University-GTB Nagar, New Delhi  
3  New Delhi     1                   Hauz Khas, New Delhi  
4  New Delhi     1             Connaught Place, New Delhi

In [60]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29753 entries, 0 to 29752
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Online_Available      29753 non-null  int64  
 1   Average_Price         29753 non-null  int64  
 2   Name                  29753 non-null  object 
 3   Menu                  29563 non-null  object 
 4   Open                  29753 non-null  int64  
 5   Cost_per_table        29753 non-null  int64  
 6   switch_to_order_menu  29753 non-null  int64  
 7   Offers                29753 non-null  object 
 8   Booking_available     29753 non-null  int64  
 9   Events                1835 non-null   object 
 10  Currency              29733 non-null  object 
 11  Rating_text           29753 non-null  object 
 12  Votes                 29753 non-null  int64  
 13  Aggregate_Rating      29753 non-null  float64
 14  Address               29753 non-null  object 
 15  City               

In [61]:
ctry = pd.read_excel("Country-Code.xlsx")
ctry

Country Code         Country
0              1           India
1             14       Australia
2             30          Brazil
3             37          Canada
4             94       Indonesia
5            148     New Zealand
6            162     Phillipines
7            166           Qatar
8            184       Singapore
9            189    South Africa
10           191       Sri Lanka
11           208          Turkey
12           214             UAE
13           215  United Kingdom
14           216   United States

In [62]:
ctry = ctry.rename(columns={'Country Code': 'C_ID'})

In [64]:
df6 = pd.merge(df5, ctry, how='left', on='C_ID')
df6

Online_Available  Average_Price                           Name  \
0                     1              3               Hauz Khas Social   
1                     0              3     Qubitos - The Terrace Cafe   
2                     1              2                The Hudson Cafe   
3                     0              3              Summer House Cafe   
4                     0              3                    38 Barracks   
...                 ...            ...                            ...   
29748                 0              1                  Senor Iguanas   
29749                 0              3  Sandpiper Restaurant & Lounge   
29750                 0              3                Texas Roadhouse   
29751                 0              1                 Riverwalk Cafe   
29752                 0              1                    Royal Hotel   

                                                    Menu  Open  \
0             Continental, American, Asian, North Indian     0   
1      Thai, European, Mexican, North Indian, Chinese...     0   
2                    Cafe, Italian, Continental, Chinese     0   
3                                   Italian, Continental     0   
4                 North Indian, Italian, Asian, American     0   
...                                                  ...   ...   
29748                                            Mexican     0   
29749                           American, Seafood, Steak     0   
29750                               American, BBQ, Steak     0   
29751                                        Asian, Thai     0   
29752                                    Pizza, Bar Food     0   

       Cost_per_table  switch_to_order_menu Offers  Booking_available  \
0                1600                     0     []                  1   
1                1500                     0     []                  1   
2                 850                     0     []                  0   
3                1850                     0     []                  1   
4                1600                     0     []                  1   
...               ...                   ...    ...                ...   
29748               0                     0     []                  0   
29749              40                     0     []                  0   
29750              45                     0     []                  0   
29751              10                     0     []                  0   
29752               0                     0     []                  0   

                                                  Events Currency Rating_text  \
0      [{'event': {'display_date': '05 April - 21 May...      Rs.   Very Good   
1      [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Excellent   
2                                                    NaN      Rs.   Very Good   
3      [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Very Good   
4                                                    NaN      Rs.   Very Good   
...                                                  ...      ...         ...   
29748                                                NaN        $        Good   
29749                                                NaN        $        Good   
29750                                                NaN        $        Good   
29751                                                NaN        $        Good   
29752                                                NaN        $        Good   

       Votes  Aggregate_Rating  \
0       7931               4.3   
1        778               4.5   
2       1537               4.4   
3       1823               4.1   
4        840               4.4   
...      ...               ...   
29748    108               3.6   
29749     85               3.6   
29750     83               3.5   
29751     91               3.6   
29752     59               3.6   

                                                 Address       City  C_ID  \
0

In [73]:

df6.head()

Online_Available  Average_Price                        Name  \
0                 1              3            Hauz Khas Social   
1                 0              3  Qubitos - The Terrace Cafe   
2                 1              2             The Hudson Cafe   
3                 0              3           Summer House Cafe   
4                 0              3                 38 Barracks   

                                                Menu  Open  Cost_per_table  \
0         Continental, American, Asian, North Indian     0            1600   
1  Thai, European, Mexican, North Indian, Chinese...     0            1500   
2                Cafe, Italian, Continental, Chinese     0             850   
3                               Italian, Continental     0            1850   
4             North Indian, Italian, Asian, American     0            1600   

   switch_to_order_menu  Offers  Booking_available  \
0                     0       0                  1   
1                     0       0                  1   
2                     0       0                  0   
3                     0       0                  1   
4                     0       0                  1   

                                              Events Currency Rating_text  \
0  [{'event': {'display_date': '05 April - 21 May...      Rs.   Very Good   
1  [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Excellent   
2                                                NaN      Rs.   Very Good   
3  [{'event': {'display_date': '16 April - 17 Apr...      Rs.   Very Good   
4                                                NaN      Rs.   Very Good   

   Votes  Aggregate_Rating                                            Address  \
0   7931               4.3             9-A & 12, Hauz Khas Village, New Delhi   
1    778               4.5  C-7, Vishal Enclave, Opposite Metro Pillar 417...   
2   1537               4.4  2524, 1st Floor, Hudson Lane, Delhi University...   
3   1823               4.1  1st Floor, DDA Shopping Complex, Aurobindo Pla...   
4    840               4.4     M-38, Outer Circle, Connaught Place, New Delhi   

        City                          Neighbourhood Country  
0  New Delhi           Hauz Khas Village, New Delhi   India  
1  New Delhi              Rajouri Garden, New Delhi   India  
2  New Delhi  Delhi University-GTB Nagar, New Delhi   India  
3  New Delhi                   Hauz Khas, New Delhi   India  
4  New Delhi             Connaught Place, New Delhi   India

In [77]:
df6['Events'] = df6['Events'].apply(lambda x: 0 if pd.isnull(x) else 1)
df6

Online_Available  Average_Price                           Name  \
0                     1              3               Hauz Khas Social   
1                     0              3     Qubitos - The Terrace Cafe   
2                     1              2                The Hudson Cafe   
3                     0              3              Summer House Cafe   
4                     0              3                    38 Barracks   
...                 ...            ...                            ...   
29748                 0              1                  Senor Iguanas   
29749                 0              3  Sandpiper Restaurant & Lounge   
29750                 0              3                Texas Roadhouse   
29751                 0              1                 Riverwalk Cafe   
29752                 0              1                    Royal Hotel   

                                                    Menu  Open  \
0             Continental, American, Asian, North Indian     0   
1      Thai, European, Mexican, North Indian, Chinese...     0   
2                    Cafe, Italian, Continental, Chinese     0   
3                                   Italian, Continental     0   
4                 North Indian, Italian, Asian, American     0   
...                                                  ...   ...   
29748                                            Mexican     0   
29749                           American, Seafood, Steak     0   
29750                               American, BBQ, Steak     0   
29751                                        Asian, Thai     0   
29752                                    Pizza, Bar Food     0   

       Cost_per_table  switch_to_order_menu  Offers  Booking_available  \
0                1600                     0       0                  1   
1                1500                     0       0                  1   
2                 850                     0       0                  0   
3                1850                     0       0                  1   
4                1600                     0       0                  1   
...               ...                   ...     ...                ...   
29748               0                     0       0                  0   
29749              40                     0       0                  0   
29750              45                     0       0                  0   
29751              10                     0       0                  0   
29752               0                     0       0                  0   

       Events Currency Rating_text  Votes  Aggregate_Rating  \
0           1      Rs.   Very Good   7931               4.3   
1           1      Rs.   Excellent    778               4.5   
2           1      Rs.   Very Good   1537               4.4   
3           1      Rs.   Very Good   1823               4.1   
4           1      Rs.   Very Good    840               4.4   
...       ...      ...         ...    ...               ...   
29748       1        $        Good    108               3.6   
29749       1        $        Good     85               3.6   
29750       1        $        Good     83               3.5   
29751       1        $        Good     91               3.6   
29752       1        $        Good     59               3.6   

                                                 Address       City  \
0                 9-A & 12, Hauz Khas Village, New Delhi  New Delhi   
1      C-7, Vishal Enclave, Opposite Metro Pillar 417...  New Delhi   
2      2524, 1st Floor, Hudson Lane, Delhi University...  New Delhi   
3      1st Floor, DDA Shopping Complex, Aurobindo Pla...  New Delhi   
4         M-38, Outer Circle, Connaught Place, New Delhi  New Delhi   
...                                                  ...        ...   
29748                 961 Hiline Rd, Pocatello, ID 83201  Pocatello   
29749                 1400 Bench Rd, Pocatello, ID 83201  Pocatello   
29750            560 Bullock Street, Pocatello, ID 8